doc/development.md

```diff
@@ -89,4 +89,205 @@
+You can lint the code before submitting a patch by doing:
+
+    make lint
+
+However, some code conventions are not explicitly enforced by the linter due
+to legacy code and inconsistent styles. But new code shall adhere to those
+rules, and old code, should be updated as possible when patched.
+
+In this section, we will cover some of the most important conventions and
+preferences for contributing to Vdsm.
+
+### Imports
+
+First import standard library modules, then 3rd party modules, and finally
+project modules.
+Separate standard library modules and 3rd party modules with a blank line.
+Keep alphabetical order within each block.
+
+Less specific imports:
+
+    from vdsm.foo import baz
+
+Should come before more specific imports like:
+
+    from vdsm.foo.bar import baz
+
+Accepted, but better to avoid:
+- Relative name imports
+- Import many modules in a comma separated list
+- Alias an imported module with non-standard abbreviation
+
+### Indentation
+
+Use 4 spaces per indentation level.
+
+Continuation lines should align wrapped elements either vertically using
+Python's implicit line joining inside parentheses, brackets and braces, or
+using a hanging indent. When using a hanging indent, the following should
+be considered:
+- There should be no arguments on the first line.
+- Further indentation should be used to clearly distinguish itself as a continuation line.
+- Hanging indentation should stick to the 4-space rule.
+```python
+# Correct
+# Add 4 spaces (an extra level of indentation) to distinguish arguments from the rest.
+foo = long_function_name(
+        var_one, var_two, var_three, var_four)
+
+# Also with if-statements.
+if (condition_1 and
+        condition_2):
+    do_something()
+
+# One line per argument for cases where arguments expand to more than one line.
+def long_function_name(
+        var_one,
+        var_two,
+        var_three,
+        var_four):
+    print(var_one)
+
+# Hanging indents should add a level.
+foo = long_function_name(
+    var_one, var_two, var_three, var_four)
+```
+```python
+# Wrong:
+# Arguments on first line forbidden (even when vertically aligned).
+foo = long_function_name(var_one, var_two,
+                         var_three, var_four)
+
+# Further indentation required as indentation is not distinguishable.
+def long_function_name(
+    var_one, var_two, var_three, var_four):
+    print(var_one)
+
+```
+
+To break comprehension Lists, Dicts, Sets as well as generator expressions,
+keep expressions that would be written in a single line together.
+Continuation lines stick to the same rules as above. Closing
+brace/bracket/parenthesis shall line up under the first character of the
+line start.
+
+```python
+# Correct
+# Vertically aligned and extra 4 spaces for condition continuation
+foo = [var for var in bar_list
+       if condition_1 and
+           condition_2]
+
+# Closing bracket starts under the first character
+foo = [
+    var for var in bar_list if condition_1
+]
```

Comment:
I wonder why the difference against function arguments.

Reply:
I don't think we have such code and there is no reason to use this formatting in this
example.

Either it fits in one line (best):

    [x for x, y in things if x]

Or you break on the condition(s):

    [x for x, y in things_that_do_not_fit()
     if x and y.startswith("foo")]

Reply:
Ok, I'll remove this example.